File: mobile/iosApp/iosApp/ViewModels/BrandingViewModel.swift
import Foundation
import SwiftUI
import shared

/**
 * ViewModel for managing tenant branding state in iOS app
 */
@MainActor
class BrandingViewModel: ObservableObject {
    @Published var branding: TenantBranding?
    @Published var isLoading: Bool = false
    @Published var errorMessage: String?

    private let brandingRepository: TenantBrandingRepository

    init(brandingRepository: TenantBrandingRepository) {
        self.brandingRepository = brandingRepository
    }

    /**
     * Load tenant branding by slug
     */
    func loadBranding(tenantSlug: String, forceRefresh: Bool = false) async {
        isLoading = true
        errorMessage = nil

        do {
            let result = try await brandingRepository.getTenantBranding(
                tenantSlug: tenantSlug,
                forceRefresh: forceRefresh
            )

            if let branding = result.getOrNull() {
                self.branding = branding
            } else if let error = result.exceptionOrNull() {
                self.errorMessage = error.localizedDescription
            }
        } catch {
            self.errorMessage = error.localizedDescription
        }

        isLoading = false
    }

    /**
     * Observe branding changes
     */
    func observeBranding(tenantSlug: String) {
        Task {
            for await result in brandingRepository.observeTenantBranding(tenantSlug: tenantSlug) {
                if let branding = result.getOrNull() {
                    self.branding = branding
                } else if let error = result.exceptionOrNull() {
                    self.errorMessage = error.localizedDescription
                }
            }
        }
    }

    /**
     * Get cached branding if available
     */
    func getCachedBranding() -> TenantBranding? {
        return brandingRepository.getCachedBranding()
    }

    /**
     * Clear branding cache
     */
    func clearCache() {
        brandingRepository.clearCache()
        branding = nil
    }

    /**
     * Get branded theme from current branding
     */
    var theme: BrandedTheme {
        return BrandedTheme(branding: branding)
    }
}

/**
 * Extension to convert KMP Result to Swift Result
 */
extension shared.Result {
    func getOrNull() -> T? {
        if self.isSuccess {
            return self.value
        }
        return nil
    }

    func exceptionOrNull() -> Error? {
        if self.isFailure {
            return self.error as? Error
        }
        return nil
    }
}

File: mobile/androidApp/src/main/kotlin/com/liyaqa/android/ui/theme/BrandingViewModel.kt
package com.liyaqa.android.ui.theme

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.liyaqa.gym.data.repositories.TenantBrandingRepository
import com.liyaqa.gym.domain.TenantBranding
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import javax.inject.Inject

/**
 * ViewModel for managing tenant branding state
 */
@HiltViewModel
class BrandingViewModel @Inject constructor(
    private val brandingRepository: TenantBrandingRepository
) : ViewModel() {

    private val _brandingState = MutableStateFlow<BrandingState>(BrandingState.Loading)
    val brandingState: StateFlow<BrandingState> = _brandingState.asStateFlow()

    /**
     * Load tenant branding by slug
     */
    fun loadBranding(tenantSlug: String, forceRefresh: Boolean = false) {
        viewModelScope.launch {
            _brandingState.value = BrandingState.Loading

            brandingRepository.getTenantBranding(tenantSlug, forceRefresh)
                .fold(
                    onSuccess = { branding ->
                        _brandingState.value = BrandingState.Success(branding)
                    },
                    onFailure = { error ->
                        _brandingState.value = BrandingState.Error(error.message ?: "Failed to load branding")
                    }
                )
        }
    }

    /**
     * Observe branding changes
     */
    fun observeBranding(tenantSlug: String) {
        viewModelScope.launch {
            brandingRepository.observeTenantBranding(tenantSlug)
                .collect { result ->
                    result.fold(
                        onSuccess = { branding ->
                            _brandingState.value = BrandingState.Success(branding)
                        },
                        onFailure = { error ->
                            _brandingState.value = BrandingState.Error(error.message ?: "Failed to load branding")
                        }
                    )
                }
        }
    }

    /**
     * Get cached branding if available
     */
    fun getCachedBranding(): TenantBranding? {
        return brandingRepository.getCachedBranding()
    }

    /**
     * Clear branding cache
     */
    fun clearCache() {
        brandingRepository.clearCache()
        _brandingState.value = BrandingState.Loading
    }
}

/**
 * Sealed class representing branding state
 */
sealed class BrandingState {
    object Loading : BrandingState()
    data class Success(val branding: TenantBranding) : BrandingState()
    data class Error(val message: String) : BrandingState()
}
